### PR TITLE
feat: Improved-error-messages-for-postgres-connection

### LINKED ISSUE #19723: [Feature]: Postgres error message improvements on connectivity
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

1. If any of the field like host address, port, DB name etc is wrong - There is a term in the error message along with the error message in the toast "**Failed to initialize pool:** ". This is not required.

2. If host address or port is wrong, there is no reference of that in the error message.

3.There is no error message when the port number is left empty.

error1
Screenshot 2023-01-11 at 1 42 11 PM

### Why should this be worked on?

1. The users should be able to see the error message wherever required.

2. The users should be able to see the right error message

### PR DESCRIPTION
**Pull request by AnnaHariprasad5123**
Hi @appsmithorg/contributor-support, @rohan-arthur, @nidhi-nair 

Fixes #19723 

**What’s in this PR:**

1. In Postgres Plugin:

- Added two imports: PSQLException and PSQLState to identify the type of exception and state of error.
- Added a condition to validate the empty port field.
- Removed the default port when the port is empty.
- Removed “Failed to initialize pool:” term in the error message.

2. In PostgresErrorMessages

- Added two constant value for Missing port and invalid host and port.

3. In PostgresPluginTest

- Added a test case to validate the missing port error message.

**Screenshots :**
1. If any of the field like host address, port, DB name etc is wrong - There is a term in the error message along with the error message in the toast "Failed to initialize pool: ". This is not required. - Fixed
image
2. If host address or port is wrong, there is no reference of that in the error message. - Fixed
image
3. There is no error message when the port number is left empty. - Fixed
image

**Test cases :**
image

**Why I didn’t provide test cases for invalid host and port :**

Those errors are triggered by the createConnectionPool method, which is private, so we can’t access it in the test file.

## Summary by CodeRabbit

- **New Features**
  - Improved validation for PostgreSQL datasource configurations, ensuring that a port is specified.
  - Enhanced error messaging for connection issues related to missing or invalid port and hostname.

- **Bug Fixes**
  - Refined error handling during connection pool initialization for clearer feedback on specific connection issues.

- **Tests**
  - Added a test to validate behavior when the datasource configuration has an empty port, improving coverage for validation logic.

### CODE REVIEW

**Comment by AnnaHariprasad5123** [comment]
Hi @nidhi-nair, Could you review this pr or assign anyone to review this pr.

**Comment by AnnaHariprasad5123** [comment]
Hi @rohan-arthur, @nidhi-nair Could you please review this pr.

**Comment by NilanshBansal** [comment]
Commenting to keep the PR alive

**Comment by NilanshBansal** [comment]
@AnnaHariprasad5123 can you resolve merge conflicts here

**Comment by AnnaHariprasad5123** [comment]
Hi @NilanshBansal, Now we can't get "Missing Port" error. Because it will take default port value. If I remove default port there, users DB will not work which we face in my previous pr.

**Comment by NilanshBansal** [comment]
> Hi @NilanshBansal, Now we can't get "Missing Port" error. Because it will take default port value. If I remove default port there, users DB will not work which we face in my previous pr.

Yes, that is fine @AnnaHariprasad5123.

**Comment by AnnaHariprasad5123** [comment]
Hi @NilanshBansal 

Good morning, I removed the code related to `Missing Port` error that fixed the cypress tests. Remaining  cypress tests are failed due to `docker.internal` url. 

**Screenshots :** 
1. cypress/e2e/Regression/ClientSide/MobileResponsiveTests/ConversionFlow_Generated_App_spec.ts
image
2. cypress/e2e/Regression/ClientSide/OneClickBinding/PropertyControl_spec.ts
image
3. cypress/e2e/Regression/ClientSide/OtherUIFeatures/EntityBottomBar_spec.ts
image
4. cypress/e2e/Smoke/GenerateCRUD/Generate_Crud_New_Page_spec.ts
image

**Comment by AnnaHariprasad5123** [comment]
Hi @NilanshBansal, Could you check this pr.

File: app/server/appsmith-plugins/postgresPlugin/src/main/java/com/external/plugins/PostgresPlugin.java [hunk 1]
```diff
@@ -1320,10 +1322,27 @@ private static HikariDataSource createConnectionPool(
...
+                            AppsmithPluginError.PLUGIN_DATASOURCE_ARGUMENT_ERROR,
+                            PostgresErrorMessages.CONNECTION_POOL_CREATION_FAILED_ERROR_MSG,
+                            cause.getMessage());
+                }
+            } else {
+                throw new AppsmithPluginException(
+                        AppsmithPluginError.PLUGIN_DATASOURCE_ARGUMENT_ERROR,
+                        PostgresErrorMessages.CONNECTION_POOL_CREATION_FAILED_ERROR_MSG,
+                        cause != null ? cause.getMessage() : e.getMessage());
+            }
```
Comment (NilanshBansal) [inline app/server/appsmith-plugins/postgresPlugin/src/main/java/com/external/plugins/PostgresPlugin.java:1345]: @AnnaHariprasad5123 can we avoid using these two else blocks to reduce code duplication as both of them return the same exception, can we just use the code structure as below.

```
if (cause instanceof PSQLException) {
    PSQLException psqlException = (PSQLException) cause;
    String sqlState = psqlException.getSQLState();
    
    if (PSQLState.CONNECTION_UNABLE_TO_CONNECT.getState().equals(sqlState)) {
        throw new AppsmithPluginException(
            AppsmithPluginError.PLUGIN_DATASOURCE_ARGUMENT_ERROR,
            PostgresErrorMessages.DS_INVALID_HOSTNAME_AND_PORT_MSG,
            psqlException.getMessage()
        );
    }
}

throw new AppsmithPluginException(
    AppsmithPluginError.PLUGIN_DATASOURCE_ARGUMENT_ERROR,
    PostgresErrorMessages.CONNECTION_POOL_CREATION_FAILED_ERROR_MSG,
    cause != null ? cause.getMessage() : e.getMessage()
);
```

**Comment by NilanshBansal** [comment]
@AnnaHariprasad5123 can you resolve merge conflicts by taking the latest pull. 
Also, there are a number of indentation changes in your last commit can you please check how these have appeared

**Comment by NilanshBansal** [comment]
@AnnaHariprasad5123 Please check the error messaging it has a full stop and a comma together. Comma should not be present
SCR-20240906-h4g

**Comment by AnnaHariprasad5123** [comment]
Hi @NilanshBansal Good morning, Could you check once now.

File: app/client/src/sagas/DatasourcesSagas.ts [hunk 2]
```diff
@@ -923,13 +923,18 @@ function* testDatasourceSaga(actionPayload: ReduxAction<Datasource>) {
...
+        });
         yield put({
           type: ReduxActionErrorTypes.TEST_DATASOURCE_ERROR,
           payload: {
+            show: false,
             id: datasource.id,
             environmentId: currentEnvironment,
-            show: true,
-            error: { message: responseData.invalids.join(", ") },
+            messages: messages,
```
Comment (NilanshBansal) [inline app/client/src/sagas/DatasourcesSagas.ts:937]: @AnnaHariprasad5123 this change would create a different toast message for each message in the invalids array.
If you want to still do toast error but without the comma,  that line can be changed to:
`show: true,
error: { message: responseData.invalids.join("\n") },`

cc: @hetunandu @sneha122

**Comment by AnnaHariprasad5123** [comment]
> @AnnaHariprasad5123 can you also check why there are a number of indentation changes in your last commit. This makes it very difficult to review. Can you please remove the unnecessary indentation changes and run `mvn spotless:apply` and then commit. image

When I save a file, the code automatically changes its formatting. This makes the lines look indented, even if I delete the extra spaces after saving it got formatted. When I run the mvn spotless:apply check command, it fails because the code doesn't match the desired formatting rules.

**Comment by NilanshBansal** [comment]
@AnnaHariprasad5123 can you check your IDE formatter rules.

**Comment by AnnaHariprasad5123** [comment]
Same config my side
image

**Comment by AnnaHariprasad5123** [comment]
@NilanshBansal, Done `mvn spotless:apply`
image

**Review by NilanshBansal — APPROVED** [review]

**Comment by NilanshBansal** [comment]
@AnnaHariprasad5123 can you please also add a unit test for this scenario. 
You can take reference from the coderabbit suggestion here
https://github.com/appsmithorg/appsmith/pull/36053#discussion_r1749803876

**Comment by AnnaHariprasad5123** [comment]
Hi @NilanshBansal, createConnectionPool method is private method. Is it good to write test case for private methods?
image

**Comment by NilanshBansal** [comment]
> Hi @NilanshBansal, createConnectionPool method is private method. Is it good to write test case for private methods? image

@AnnaHariprasad5123 thanks for highlighting we can skip it in this scenario.

**Pull request merged**